ipa-core/src/helpers/gateway/mod.rs

```diff
@@ -6,41 +6,52 @@ mod transport;
 
 use std::num::NonZeroUsize;
 
-pub(super) use receive::ReceivingEnd;
+pub(super) use receive::{MpcReceivingEnd, ShardReceivingEnd};
 pub(super) use send::SendingEnd;
-#[cfg(all(test, feature = "shuttle"))]
-use shuttle::future as tokio;
 #[cfg(feature = "stall-detection")]
 pub(super) use stall_detection::InstrumentedGateway;
+pub use transport::RoleResolvingTransport;
 
 use crate::{
     helpers::{
         buffers::UnorderedReceiver,
         gateway::{
-            receive::GatewayReceivers, send::GatewaySenders, transport::RoleResolvingTransport,
+            receive::{GatewayReceivers, ShardReceiveStream, UR},
+            send::GatewaySenders,
+            transport::Transports,
         },
-        transport::routing::RouteId,
-        HelperChannelId, LogErrors, Message, Role, RoleAssignment, TotalRecords, Transport,
+        HelperChannelId, LogErrors, Message, MpcMessage, RecordsStream, Role, RoleAssignment,
+        ShardChannelId, TotalRecords, Transport,
     },
     protocol::QueryId,
+    sharding::ShardIndex,
+    sync::{Arc, Mutex},
 };
 
 /// Alias for the currently configured transport.
 ///
 /// To avoid proliferation of type parameters, most code references this concrete type alias, rather
 /// than a type parameter `T: Transport`.
 #[cfg(feature = "in-memory-infra")]
-pub type TransportImpl = super::transport::InMemoryTransport<crate::helpers::HelperIdentity>;
+type TransportImpl<I> = super::transport::InMemoryTransport<I>;
+#[cfg(feature = "in-memory-infra")]
+pub type MpcTransportImpl = TransportImpl<crate::helpers::HelperIdentity>;
+#[cfg(feature = "in-memory-infra")]
+pub type ShardTransportImpl = TransportImpl<ShardIndex>;
 
 #[cfg(feature = "real-world-infra")]
-pub type TransportImpl = crate::sync::Arc<crate::net::HttpTransport>;
+type TransportImpl = crate::sync::Arc<crate::net::HttpTransport>;
+#[cfg(feature = "real-world-infra")]
+pub type MpcTransportImpl = TransportImpl;
+#[cfg(feature = "real-world-infra")]
+pub type ShardTransportImpl = crate::net::HttpShardTransport;
```

Comment:
```suggestion
#[cfg(feature = "real-world-infra")]
pub type MpcTransportImpl = crate::sync::Arc<crate::net::HttpTransport>;
#[cfg(feature = "real-world-infra")]
pub type ShardTransportImpl = crate::net::HttpShardTransport;
```

I think better not to give the impression these two types have a common interface. (I'm not sure the intermediate `TransportImpl` alias has much value in the in-memory case either, vs. importing `InMemoryTransport`, but that's a more minor issue.)